### PR TITLE
#419 Refactor guice config for Server

### PR DESCRIPTION
**Pull request by kevinconaway**
Fixes for #419 

Jackson doesn't seem to properly handle generics so it won't find any `TypeLiteral` bindings.  I refactored this to use an explicit binding with a @Named parameter.

The `JmxTransfomerIT` test wasn't running as part of the build (and would have caught this) so I renamed it to to actually run during the build.
<a href='#crh-start'></a><a href='#crh-data-%7B%22processed%22%3A%20%5B%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51572466%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23issuecomment-178777648%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23issuecomment-178812723%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51637222%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51637316%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23issuecomment-178831625%22%2C%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51696035%22%5D%2C%20%22comments%22%3A%20%7B%22General%20Comment%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23issuecomment-178777648%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Looks%20like%20the%20CloudWatchWriterIT%20is%20not%20compatible%20with%20the%20LessIOSecurityManager%22%2C%20%22created_at%22%3A%20%222016-02-02T19%3A34%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6921953%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kevinconaway%22%7D%7D%2C%20%7B%22body%22%3A%20%22%3E%20Looks%20like%20the%20CloudWatchWriterIT%20is%20not%20compatible%20with%20the%20LessIOSecurityManager%5Cr%5Cn%5Cr%5Cn%40gehel%20what%20are%20your%20thoughts%20here%3F%20%20The%20LessIOSecurityManager%20doesn%27t%20seem%20to%20play%20nice%20with%20Powermock%20%28javassist%29%20or%20custom%20test%20runners%20%28either%20the%20PowerMockRunner%20or%20MockitoJUnitRunner%29.%5Cr%5Cn%5Cr%5CnOne%20solution%20that%20I%20prototyped%20was%20to%20allow%20Guice%20in%20the%20OutputWriters%20so%20that%20we%20could%20properly%20use%20DI%2C%20instead%20of%20doing%20things%20like%20creating%20the%20%60AmazonCloudWatchClient%60%20directly%20in%20code.%5Cr%5Cn%5Cr%5CnMy%20thought%20was%20we%20could%20leverage%20the%20java%20ServiceProvider%20infrastructure%20to%20allow%20the%20different%20output%20writers%20%28or%20really%20any%20maven%20submodule%20component%29%20to%20contribute%20Guice%20modules.%5Cr%5Cn%5Cr%5CnEach%20sub-module%20could%20optionally%20implement%20the%20following%20SPI%20interface%3A%5Cr%5Cn%5Cr%5Cn%60%60%60java%5Cr%5Cnimport%20com.google.inject.Module%3B%5Cr%5Cn%5Cr%5Cnpublic%20interface%20ModuleProvider%20%7B%5Cr%5Cn%5Cr%5Cn%5CtModule%20getModule%28%29%3B%5Cr%5Cn%5Cr%5Cn%7D%5Cr%5Cn%60%60%60%5Cr%5Cn%5Cr%5CnThe%20modules%20would%20then%20be%20loaded%20from%20the%20ServiceLoader%3A%5Cr%5Cn%5Cr%5Cn%60%60%60java%5Cr%5Cn%40Nonnull%5Cr%5Cnpublic%20static%20Injector%20createInjector%28%40Nonnull%20JmxTransConfiguration%20configuration%29%20%7B%5Cr%5Cn%5Ctreturn%20Guice.createInjector%28%5Cr%5Cn%5Ct%5CtcreateModules%28configuration%29%5Cr%5Cn%5Ct%29%3B%5Cr%5Cn%7D%5Cr%5Cn%5Cr%5Cnpublic%20static%20Collection%3CModule%3E%20createModules%28%40Nonnull%20JmxTransConfiguration%20configuration%29%20%7B%5Cr%5Cn%5CtList%3CModule%3E%20modules%20%3D%20new%20ArrayList%3CModule%3E%28%29%3B%5Cr%5Cn%5Cr%5Cn%5Ctmodules.add%28new%20JmxTransModule%28configuration%29%29%3B%5Cr%5Cn%5Ctmodules.add%28%5Cr%5Cn%5Ct%5Ctnew%20ObjectMapperModule%28%29%5Cr%5Cn%5Ct%5Ct%5Ct.registerModule%28new%20GuavaModule%28%29%29%5Cr%5Cn%5Ct%29%3B%5Cr%5Cn%5Cr%5Cn%5Ctfor%20%28ModuleProvider%20moduleProvider%20%3A%20ServiceLoader.load%28ModuleProvider.class%29%29%20%7B%5Cr%5Cn%5Ct%5Ctmodules.add%28moduleProvider.getModule%28%29%29%3B%5Cr%5Cn%5Ct%7D%5Cr%5Cn%5Cr%5Cn%5Ctreturn%20modules%3B%5Cr%5Cn%7D%5Cr%5Cn%60%60%60%5Cr%5Cn%5Cr%5CnThoughts%3F%22%2C%20%22created_at%22%3A%20%222016-02-02T20%3A49%3A44Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6921953%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kevinconaway%22%7D%7D%2C%20%7B%22body%22%3A%20%22Adding%20a%20SPI%20integration%20is%20probably%20for%20another%20PR.%20The%20issue%20with%20PowerMock%20should%20be%20solved%20by%20removing%20PowerMock.%20The%20fact%20that%20we%20need%20PowerMock%20is%20a%20sign%20of%20a%20bad%20design.%5Cr%5Cn%5Cr%5CnNot%20having%20LessIOSecurityManager%20for%20integration%20tests%20is%20fine.%20We%20do%20expect%20integration%20tests%20to%20have%20IO.%5Cr%5Cn%5Cr%5CnI%20don%27t%20really%20have%20any%20issue%20with%20%60AmazonCloudWatchClient%60%20being%20created%20by%20code.%20It%20is%20more%20or%20less%20nicely%20isolated%20in%20its%20own%20factory.%20I%20would%20have%20much%20more%20concern%20having%20it%20in%20the%20global%20Guice%20scope.%20It%20is%20used%20only%20by%20a%20specific%20module%2C%20it%20should%20not%20be%20accessible%20from%20outside%20of%20it.%20So%20let%27s%20say%20that%20I%20don%27t%20really%20see%20a%20reason%20to%20have%20an%20SPI%20integration.%5Cr%5Cn%5Cr%5CnI%27m%20happy%20to%20continue%20this%20conversation.%20You%20might%20be%20able%20to%20convince%20me%20...%22%2C%20%22created_at%22%3A%20%222016-02-02T21%3A29%3A32Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22General%20Comment%22%7D%2C%20%22Pull%20b7fdc395ef905c492e52b9c86660a821520527fb%20pom.xml%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51637222%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22Are%20you%20sure%20about%20removing%20this%20dependency%3F%22%2C%20%22created_at%22%3A%20%222016-02-02T21%3A22%3A50Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%2C%20%7B%22body%22%3A%20%22It%20is%20a%20duplicate%2C%20see%20the%20one%20directly%20above%20it.%22%2C%20%22created_at%22%3A%20%222016-02-02T21%3A23%3A21Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/6921953%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/kevinconaway%22%7D%7D%2C%20%7B%22body%22%3A%20%22My%20bad%20...%22%2C%20%22created_at%22%3A%20%222016-02-03T09%3A23%3A16Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20pom.xml%3AL461-467%22%7D%2C%20%22Pull%2076a30dc69bc882268aa7321dfc72e76f55f21dde%20jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java%204%22%3A%20%7B%22html_url%22%3A%20%22https%3A//github.com/jmxtrans/jmxtrans/pull/420%23discussion_r51572466%22%2C%20%22comments%22%3A%20%5B%7B%22body%22%3A%20%22The%20correct%20fix%20should%20be%20to%20configure%20%5BMaven%20Failsafe%20plugin%5D%28https%3A//maven.apache.org/surefire/maven-failsafe-plugin/%29.%20I%27d%20like%20to%20make%20sure%20that%20unit%20tests%20are%20actually%20testing%20small%20units%20and%20are%20fast.%20As%20we%20are%20also%20using%20%5BPIT%5D%28http%3A//pitest.org/%29%2C%20it%20is%20specially%20important%20that%20unit%20tests%20are%20fast.%5Cr%5Cn%5Cr%5CnCould%20you%20rollback%20this%20specific%20change%20and%20open%20an%20issue%20to%20add%20an%20integration%20test%20phase%20to%20the%20build%20%3F%22%2C%20%22created_at%22%3A%20%222016-02-02T14%3A13%3A56Z%22%2C%20%22user%22%3A%20%7B%22avatar_url%22%3A%20%22https%3A//avatars.githubusercontent.com/u/1415765%3Fv%3D3%22%2C%20%22html_url%22%3A%20%22https%3A//github.com/gehel%22%7D%7D%5D%2C%20%22title%22%3A%20%22File%3A%20jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java%3AL46-54%22%7D%7D%7D'></a>
<a href='https://www.codereviewhub.com/'></a>
- [x] <a href='#crh-comment-Pull 76a30dc69bc882268aa7321dfc72e76f55f21dde jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java 4'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/420#discussion_r51572466'>File: jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java:L46-54</a></b>
- <a href='https://github.com/gehel'></a> The correct fix should be to configure [Maven Failsafe plugin](https://maven.apache.org/surefire/maven-failsafe-plugin/). I'd like to make sure that unit tests are actually testing small units and are fast. As we are also using [PIT](http://pitest.org/), it is specially important that unit tests are fast.
Could you rollback this specific change and open an issue to add an integration test phase to the build ?
- [ ] <a href='#crh-comment-General Comment'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/420#issuecomment-178777648'>General Comment</a></b>
- <a href='https://github.com/kevinconaway'></a> Looks like the CloudWatchWriterIT is not compatible with the LessIOSecurityManager
- <a href='https://github.com/kevinconaway'></a> > Looks like the CloudWatchWriterIT is not compatible with the LessIOSecurityManager
@gehel what are your thoughts here?  The LessIOSecurityManager doesn't seem to play nice with Powermock (javassist) or custom test runners (either the PowerMockRunner or MockitoJUnitRunner).
One solution that I prototyped was to allow Guice in the OutputWriters so that we could properly use DI, instead of doing things like creating the `AmazonCloudWatchClient` directly in code.
My thought was we could leverage the java ServiceProvider infrastructure to allow the different output writers (or really any maven submodule component) to contribute Guice modules.
Each sub-module could optionally implement the following SPI interface:
```java
import com.google.inject.Module;
public interface ModuleProvider {
Module getModule();
}
```
The modules would then be loaded from the ServiceLoader:
```java
@Nonnull
public static Injector createInjector(@Nonnull JmxTransConfiguration configuration) {
return Guice.createInjector(
createModules(configuration)
);
}
public static Collection<Module> createModules(@Nonnull JmxTransConfiguration configuration) {
List<Module> modules = new ArrayList<Module>();
modules.add(new JmxTransModule(configuration));
modules.add(
new ObjectMapperModule()
.registerModule(new GuavaModule())
);
for (ModuleProvider moduleProvider : ServiceLoader.load(ModuleProvider.class)) {
modules.add(moduleProvider.getModule());
}
return modules;
}
```
Thoughts?
- <a href='https://github.com/gehel'></a> Adding a SPI integration is probably for another PR. The issue with PowerMock should be solved by removing PowerMock. The fact that we need PowerMock is a sign of a bad design.
Not having LessIOSecurityManager for integration tests is fine. We do expect integration tests to have IO.
I don't really have any issue with `AmazonCloudWatchClient` being created by code. It is more or less nicely isolated in its own factory. I would have much more concern having it in the global Guice scope. It is used only by a specific module, it should not be accessible from outside of it. So let's say that I don't really see a reason to have an SPI integration.
I'm happy to continue this conversation. You might be able to convince me ...
- [ ] <a href='#crh-comment-Pull b7fdc395ef905c492e52b9c86660a821520527fb pom.xml 4'></a> &nbsp;<b><a href='https://github.com/jmxtrans/jmxtrans/pull/420#discussion_r51637222'>File: pom.xml:L461-467</a></b>
- <a href='https://github.com/gehel'></a> Are you sure about removing this dependency?
- <a href='https://github.com/kevinconaway'></a> It is a duplicate, see the one directly above it.
- <a href='https://github.com/gehel'></a> My bad ...

<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/420?mark_as_completed=1'></a>&nbsp;<a href='https://www.codereviewhub.com/jmxtrans/jmxtrans/pull/420?approve=1'></a>&nbsp;<a href='https://github.com/jmxtrans/jmxtrans/pull/420'></a>
<a href='#crh-end'></a>

### CODE REVIEW

**Comment by gehel** [comment]
This pull request is now managed with CodeReviewHub. <a href='https://github.com/jmxtrans/jmxtrans/pull/420'>Refresh</a>.<a href='#crh-update-none'></a>

File: jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java [hunk 1]
```diff
@@ -46,8 +46,8 @@
 
 @AllowDNSResolution
 @AllowLocalFileAccess(paths = "*")
-@Category({IntegrationTest.class, RequiresIO.class})
```
Comment (gehel) [inline jmxtrans-core/src/test/java/com/googlecode/jmxtrans/JmxTransformerTest.java]: The correct fix should be to configure [Maven Failsafe plugin](https://maven.apache.org/surefire/maven-failsafe-plugin/). I'd like to make sure that unit tests are actually testing small units and are fast. As we are also using [PIT](http://pitest.org/), it is specially important that unit tests are fast.

Could you rollback this specific change and open an issue to add an integration test phase to the build ?

**Comment by kevinconaway** [comment]
Looks like the CloudWatchWriterIT is not compatible with the LessIOSecurityManager

**Comment by kevinconaway** [comment]
> Looks like the CloudWatchWriterIT is not compatible with the LessIOSecurityManager

@gehel what are your thoughts here?  The LessIOSecurityManager doesn't seem to play nice with Powermock (javassist) or custom test runners (either the PowerMockRunner or MockitoJUnitRunner).

One solution that I prototyped was to allow Guice in the OutputWriters so that we could properly use DI, instead of doing things like creating the `AmazonCloudWatchClient` directly in code.

My thought was we could leverage the java ServiceProvider infrastructure to allow the different output writers (or really any maven submodule component) to contribute Guice modules.

Each sub-module could optionally implement the following SPI interface:

```java
import com.google.inject.Module;

public interface ModuleProvider {

	Module getModule();

}
```

The modules would then be loaded from the ServiceLoader:

```java
@Nonnull
public static Injector createInjector(@Nonnull JmxTransConfiguration configuration) {
	return Guice.createInjector(
		createModules(configuration)
	);
}

public static Collection<Module> createModules(@Nonnull JmxTransConfiguration configuration) {
	List<Module> modules = new ArrayList<Module>();

	modules.add(new JmxTransModule(configuration));
	modules.add(
		new ObjectMapperModule()
			.registerModule(new GuavaModule())
	);

	for (ModuleProvider moduleProvider : ServiceLoader.load(ModuleProvider.class)) {
		modules.add(moduleProvider.getModule());
	}

	return modules;
}
```

Thoughts?

File: pom.xml [hunk 2]
```diff
@@ -461,12 +461,6 @@
 				<scope>runtime</scope>
 			</dependency>
 			<dependency>
-				<groupId>org.jboss.remoting3</groupId>
```
Comment (gehel) [inline pom.xml]: Are you sure about removing this dependency?

Comment (kevinconaway) [inline pom.xml]: It is a duplicate, see the one directly above it.

**Comment by gehel** [comment]
Adding a SPI integration is probably for another PR. The issue with PowerMock should be solved by removing PowerMock. The fact that we need PowerMock is a sign of a bad design.

Not having LessIOSecurityManager for integration tests is fine. We do expect integration tests to have IO.

I don't really have any issue with `AmazonCloudWatchClient` being created by code. It is more or less nicely isolated in its own factory. I would have much more concern having it in the global Guice scope. It is used only by a specific module, it should not be accessible from outside of it. So let's say that I don't really see a reason to have an SPI integration.

I'm happy to continue this conversation. You might be able to convince me ...

File: pom.xml (hunk 2 above)
Comment (gehel) [inline pom.xml]: My bad ...

**Pull request merged**